### PR TITLE
feat: migrate project to TypeScript

### LINKED ISSUE #26: Migrate to typescript
Thinking about rewriting current codebase with TS, cause there is no so much code right now, it mustn't take much time with it

### PR DESCRIPTION
**Pull request by mxschmitt**
TODOs:
- [x] most of the type any needs to be fixed
- [x] figure out why we can't run the tests all together, probably forgot to unmock something -> #46
- [x] @mmarkelov please enable auto deletion of merged branches in the repository settings. This will keep the branches clean
- [x] maybe add codecov? cc @mmarkelov -> #46

Closes #26

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export const FIREFOX = 'firefox'
+export const WEBKIT = 'webkit'
+
+export const PARALLEL = '--parallel'
+
+export interface Config {
+  launchBrowserApp?: LaunchOptions
+  context?: BrowserContextOptions,
+  exitOnPageError: boolean,
+  browser?: string
```
Comment (mmarkelov) [inline src/constants.ts]: I suppose it should be one of **CHROMIUM**, **FIREFOX** or **WEBKIT**

File: package.json [hunk 2]
```diff
@@ -23,13 +23,11 @@
   },
   "homepage": "https://github.com/mmarkelov/jest-playwright#readme",
   "scripts": {
-    "build": "rm -rf lib/ && babel src -d lib --ignore '**/*.test.js'",
+    "build": "tsc",
     "format": "prettier --write \"*.{js,md}\" \"src/**/*.js\"",
     "lint": "eslint .",
     "prepublishOnly": "yarn run build",
-    "test:bin": "jest src/bin/utils.test.js",
```
Comment (mxschmitt) [inline package.json]: I have not understood why you splitted them up there. So for each a new Jest instance would launch which would result in a longer test run.

**Comment by mmarkelov** [comment]
@mxschmitt this looks really great! Good job. I suppose we should fix warnings and we can merge it.  Also I just send you collab invite, cause sometimes I couldn't answer quickly.

**Comment by mxschmitt** [comment]
Okay, I think we are mostly ready now for a final review. :)
@mmarkelov

File: src/bin/utils.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { checkBrowserEnv } from '../utils'
 
 /* eslint import/prefer-default-export:0 */
-export function checkBrowsers(browsers) {
+export function checkBrowsers(browsers?: string[]) {
```
Comment (mmarkelov) [inline src/bin/utils.ts]: it should be ```browsers?: BrowserType[]``` I suppose

File: src/bin/testProcess.ts [hunk 4]
```diff
@@ -12,7 +12,11 @@ const getSpawnOptions = browser => ({
   },
 })
 
-const exec = ({ sequence, browser, params }) => {
+const exec = ({ sequence, browser, params }: {
+  sequence: string,
+  browser: string,
```
Comment (mmarkelov) [inline src/bin/testProcess.ts]: ```BrowserType``` instead of ```string```

File: src/bin/testProcess.ts [hunk 5]
```diff
@@ -1,9 +1,9 @@
-import { spawn, spawnSync } from 'child_process'
+import { spawn, spawnSync, SpawnSyncOptions } from 'child_process'
 import { readConfig } from '../utils'
 import { checkBrowsers } from './utils'
 import { PARALLEL } from '../constants'
 
-const getSpawnOptions = browser => ({
+const getSpawnOptions = (browser: string): SpawnSyncOptions => ({
```
Comment (mmarkelov) [inline src/bin/testProcess.ts]: And here ```string``` -> ```BrowserType```

File: package.json [hunk 6]
```diff
@@ -23,13 +23,11 @@
...
-    "test:utils": "jest src/utils.test.js",
-    "test": "npm run test:utils && npm run test:bin && jest tests"
+    "test": "jest src/bin/utils.test.ts tests/stub.test.ts && jest src/utils.test.ts"
```
Comment (mmarkelov) [inline package.json]: Shouldn't it be ```jest src/bin/utils.test.ts tests/stub.test.ts src/utils.test.ts```? To be more consistence

Comment (mxschmitt) [reply package.json]: By having all in the same jest statement, the run will fail. I've created a separate issue to fix this minor issue later: #46.

File: src/utils.ts [hunk 7]
```diff
@@ -64,7 +64,7 @@ export async function readPackage() {
   return playwright
 }
 
-export async function getPlaywrightInstance(browserType) {
+export async function getPlaywrightInstance(browserType: string) {
```
Comment (mmarkelov) [inline src/utils.ts]: shouldn't it be ```BrowserType```?

File: package.json (hunk 6 above)
Comment (mmarkelov) [reply package.json]: @mxschmitt got it!

File: src/utils.ts (hunk 7 above)
Comment (mxschmitt) [reply src/utils.ts]: true 👍

**Comment by mmarkelov** [comment]
@mxschmitt i think that we can merge it

**Pull request merged**